designs/build_layer_support.rst

```diff
@@ -0,0 +1,222 @@
+      Runtime.
+* Building for runtimes that ``sam build`` command currently does not support
+* Building Layers that define no compatible runtimes
+
+
+User Experience Walkthrough
+---------------------------
+Experience is the same as the initially design of `build <sam_build_cmd.rst>`__ with that addition of supporting
+that workflow with templates that have Layer Resources.
+
+A Layer Resource's Content location (local directory) must have a manifest file. This is the same restrictions Functions
+have and we will keep these consistent across Layers and Functions.
+
+Implementation
+==============
+
+CLI Changes
+-----------
+*Explain the changes to command line interface, including adding new commands, modifying arguments etc*
+
+No changes to the CLI's interface needed to support building Layers.
+
+Breaking Change
+~~~~~~~~~~~~~~~
+*Are there any breaking changes to CLI interface? Explain*
+
+No
+
+Design
+------
+*Explain how this feature will be implemented. Highlight the components of your implementation, relationships*
+*between components, constraints, etc.*
+
+The BuildContext currently has a property for a function_provider. As we expand to building more resources, we need
+the framework to support this seamlessly.
+
+.. code-block:: python
+    class BuildContext(object):
+
+        ...
+
+        @property
+        def function_provider(self):
+            return self._function_provider
+
+        ...
+
+
+We will update this the ``function_provider`` property to be a ``resource_provider`` that will return the list of
+providers that will be used to get different types of resources from the template.
+
+
+.. code-block:: python
+    class BuildContext(object):
+
+        ...
+
+        @property
+        def resource_providers(self):
+            return [self._function_provider, self._layer_provider]
+
+        ...
+
+
+With this, we need to expand the ApplicationBuilder class as well. Since we are building different resources. The class
+will be updated to:
+
+.. code-block:: python
+    class ApplicationBuilder(object):
+
+        def __init__(self,
+                     provider_list,
+                     build_dir,
+                     base_dir,
+                     manifest_path_override=None,
+                     container_manager=None,
+                     parallel=False):
+            pass
+
+        def build(self):
+            """
+            Build the entire application.
+
+            For each provider in provider_list:
+                For each resource in the provider:
+                    build
+
+            Returns
+            -------
+            dict
+                Returns the path to where each resource was built as a map of resource's LogicalId to the path string
+            """
+            pass
+
+
+The ApplicationBuilder no longer understands a single provider. This requires the building logic to live at each
```

Comment:
Will this result in any partial builds happening? as in we have 4 layers, if they are built independently and fail on the 4th, do we still have the other artifacts built?

Reply:
I didn't think of this case. Great call out. 

My initial thought is: This should be no different than functions. If you have 4 functions and one fails what happens (I don't know the answer to this yet but will test it out). On the other hand, if you are embedding Layers in with your template/functions (Ref'ing a Layer resource in the Function definition). You will want your function invokable only if all the layers you are referencing have also built. 

I am going to ponder this a little more to see which one is better or maybe how we can support both.